File: two_pointers/minimum_window_sort.py


# minimum_window_sort
# length of the smallest subsequence, when sorted will sort the whole array
# two pointers approach
# O(n) time complexity, O(1) space
# step 1 - traverse back/forth from each end of the sequence to find sorted part of the sequence
def minimum_window_sort(seq):

    if seq is None:
        return

    ascending = seq[-1] > seq[0]

    p = 0
    q = len(seq)-1
    while p < q:

        if ascending:
            if seq[p] > seq[p+1] and seq[q] < seq[q-1]:
                return seq[p:q+1]

            if seq[p+1] >= seq[p]:
                p = p + 1
            if seq[q-1] <= seq[q]:
                q = q - 1
        else:
            if seq[p] < seq[p+1] and seq[q] > seq[q-1]:
                return seq[p:q+1]

            if seq[p+1] <= seq[p]:
                p = p + 1
            if seq[q-1] >= seq[q]:
                q = q - 1

    return seq[p:q+1]


def main():
    seq = [1, 2, 5, 3, 7, 10, 9, 12]
    res = minimum_window_sort(seq)
    print(f'{res}')

    seq2 = [1, 2, 3, 5, 8, 7, 9, 12]
    res = minimum_window_sort(seq2)
    print(f'{res}')

    seq3 = [1, 2, 3, 5, 7, 7, 9, 12]
    res = minimum_window_sort(seq3)
    print(f'{res}')

    seq4 = [12,11,9,3,4,1]
    res = minimum_window_sort(seq4)
    print(f'{res}')


if __name__ == "__main__":
    main()

File: twopointers/minimumwindowsort.go
package twopointers

// mimimumwindowsort
// find minumum window when sorted will sort the whole slice
// two pointers approach
// O(n) time and O(1) space
func MinimumWindowSort(s []int) []int {

	if len(s) == 0 {
		return make([]int, 0)
	}

	p := 0
	q := len(s)-1
	for ;p < q; {

		if s[p] > s[p+1] && s[q] < s[q-1] {
			break
		}
		if s[p] < s[p+1] {
			p += 1
		}
		if s[q] > s[q-1] {
			q -= 1
		}
	}

	return s[p+1:q+1]
}
